File: 800/longest_palindrome.cpp
#include <bits/stdc++.h>

using namespace std;

int main()
{
    ios::sync_with_stdio(0);
    cin.tie(0);

    int n, m;
    cin >> n >> m;

    set<string> all_strings;
    for (int i = 0; i < n; i++) {
        string s;
        cin >> s;
        all_strings.insert(s);
    }

    vector<string> start;
    vector<string> end;
    string individual_string = "";
    for (string s : all_strings) {
        string rev = s;
        reverse(rev.begin(), rev.end());
        if (all_strings.find(rev) != all_strings.end() && find(start.begin(), start.end(), s) == start.end() && find(end.begin(), end.end(), s) == end.end() && s != rev) {
            start.push_back(s);
            end.push_back(rev);
        } else if (s == rev) {
            individual_string = s;
        }
    }

    string answer = "";
    for (string s : start) {
        answer.append(s);
    }
    answer.append(individual_string);
    reverse(end.begin(), end.end());
    for (string s : end) {
        answer.append(s);
    }

    cout << answer.length() << endl;
    cout << answer << endl;
}

File: 800/longest_palindrome.py
n, m = [int(x) for x in input().split()]

all_strings = set()
for i in range(n):
    all_strings.add(input())

start = []
end = []
individual_string = ""
for s in all_strings:
    if s[::-1] in all_strings and s not in start and s not in end and s != s[::-1]:
        start.append(s)
        end.append(s[::-1])
    elif s == s[::-1]:
        individual_string = s

start = "".join(start)
end = "".join(reversed(end))
answer = start + individual_string + end
print(len(answer))
print(answer)
